Replace fully-automated release process with staged release process

phargogh opened the pull request:
This PR replaces our current fully-automated release process with one that instead requires us to make our own decisions about whether a release should go out.

There are 2 components of this:
* The [simplified bugfix release checklist,](https://github.com/natcap/invest/wiki/Bugfix-Release-Checklist) distilled down to the actions we need to take as a team.
* A series of scripts to do the boring, trivial stuff (updating HISTORY, updating the Makefile with the latest UG repo rev, committing to a new branch, etc.).

All parts of the release _can_ be done by hand (and steps are described in the [release checklist](https://github.com/natcap/invest/wiki/Bugfix-Release-Checklist)), but the series of scripts will help automate this if desired:

```shell
$ ./scripts/release-1-initialize.sh 3.8.8
HISTORY has been updated to release version 3.8.8 on 2020-07-23

Changes have been made to the following files:
  * HISTORY.rst has been updated with the release version and today's date
  * Makefile has been updated with the latest user's guide revision

To continue with the release:
  $ ./scripts/release-2-commit.sh 3.8.8
```

Changes can be committed by following the [steps in the checklist](https://github.com/natcap/invest/wiki/Bugfix-Release-Checklist#commit-changes-for-release) or by running the helper script:
```shell
$ ./scripts/release-2-commit.sh 3.8.8
M	HISTORY.rst
M	Makefile
Switched to a new branch 'autorelease/3.8.8'
[autorelease/3.8.8 c8a067bd] Committing the 3.8.8 release.
 2 files changed, 7 insertions(+), 3 deletions(-)
The release has been committed and tagged.

To push to natcap/invest:
  $ git push git@github.com:natcap/invest.git 3.8.8 autorelease/3.8.8

To push to your fork:
  $ git push git@github.com:phargogh/invest.git 3.8.8 autorelease/3.8.8

After pushing, wait for builds to finish before continuing.
    See https://github.com/natcap/invest/wiki/Bugfix-Release-Checklist#wait-for-builds-to-complete
```

The final step of this is to do publish the release to GitHub and PyPI (artifacts are already pushed to GCS).  Again, this is [documented in the release checklist](https://github.com/natcap/invest/wiki/Bugfix-Release-Checklist#publish-the-release), or the script can be run to automate the various steps (requires `gsutil`, `hub`, `pandoc`, `envsubst`, `twine`):

```shell
$ ./scripts/release-3-publish.sh
```

Since this will affect us all on the software team it'd be great to have everyone's input on this process.  Thanks in advance!

dcdenu4 commented on .github/workflows/build-and-test.yml:
```diff
@@ -141,6 +141,28 @@ jobs:
+                  $env:GSUTIL_PATH = (Get-Command gsutil).path
+
+                  # Specify which python version we want to use (it's the one we're using
+                  # in actions/setup-python.
```
Closing paren here for the comment?

phargogh replied:
Patched in 1d205d67

dcdenu4 commented on .github/workflows/build-and-test.yml:
```diff
@@ -200,6 +222,32 @@ jobs:
+                  $env:GSUTIL_PATH = (Get-Command gsutil).path
+
+                  # Specify which python version we want to use (it's the one we're using
+                  # in actions/setup-python.
```
Same closing paren comment here.

phargogh replied:
Patched in 774cf4e0

dcdenu4 commented on scripts/release-2-commit.sh:
```diff
@@ -0,0 +1,32 @@
+VERSION=$1
+: "${VERSION:?'The version string is needed as parameter 1.'}"
+
+set -e
```
This exits the script on first error?

phargogh replied:
Yep!  I added a brief comment about this in bd03b586 as well, for future us.

dcdenu4 commented on scripts/release-3-publish.sh:
```diff
@@ -0,0 +1,48 @@
+    "$VERSION"
+
+PR_MSG_TEXT=build/pr_msg_text_$VERSION.txt
+SOURCE_BRANCH="master"  # TODO: does this need to be exported?
+BUGFIX_VERSION="$VERSION"  # TODO: does this need to be exported?
```
Noticed a couple TODO comments here. Are these still active questions?

phargogh replied:
Ah, thanks for the reminder about this!  I had forgotten to implement the fix.  The best answer here, I think, is not to `export` the variables but instead provide them directly to the subshell that runs `envsubst` where they're needed.  That way we don't unnecessarily pollute the parent shell with extraneous environment variables.  Kind of a neat feature of bash-like shells!  Patched in 4e6d4cdf

dcdenu4 commented on Makefile:
```diff
@@ -82,12 +82,14 @@ BUILD_DIR := build
 # a zipfile of the source code).
 FORKNAME := $(word 2, $(subst :,,$(subst github.com, ,$(shell git remote get-url origin))))
 FORKUSER := $(word 1, $(subst /, ,$(FORKNAME)))
+RELEASES_BUCKET := gs://releases.naturalcapitalproject.org
+DEV_BUILD_BUCKET := gs://natcap-dev-build-artifacts
```
Were these just moved to their own variables for readability and style? I couldn't see if these variables end up being used elsewhere...

phargogh replied:
Yeah, I separated these out in `Makefile` so that they can be referred to directly in `scripts/release-3-publish.sh`:

```shell
# assume a default user of "natcap" if none has been provided.
GH_USER="${2:-'natcap'}"
if [ "$GH_USER" = "natcap" ]
then
    BUCKET="$(make jprint-RELEASES_BUCKET)/invest"
else
    BUCKET="$(make jprint-DEV_BUILD_BUCKET)/invest/$GH_USER"
fi
```

I added a comment to `Makefile` in 10f9df17

dcdenu4 replied:
Ah I see, thanks.

dcdenu4 commented on .github/workflows/build-and-test.yml:
```diff
@@ -141,6 +141,28 @@ jobs:
                   name: Wheel for ${{ matrix.os }} ${{ matrix.python-version }} ${{ matrix.python-architecture }}
                   path: dist
 
+            - name: Set up GCP
```
Could you explain briefly the GCP / GCS code here and when this will be run? Are artifacts deployed only on push to master?

phargogh replied:
Artifacts will be deployed on any push to any branch.  So on `natcap/invest` the deployment would happen whenever a PR into `master` or `release/*` is merged, or when a release branch is pushed as part of the release process.  If you have the right secret defined in your fork, the deployment would happen on each push there, too.

The reason we have the `if: github.event_name != 'pull_request'` condition is that deploying to GCS requires the service account key to be defined and secrets aren't provided to actions run within PRs (it'd be a pretty glaring security risk if they were!)

dcdenu4 replied:
Thanks!

phargogh commented:
Thanks @dcdenu4 !  I addressed your comments and pushed a few more changes, mostly with a few comments, but also with an improvement on the passing of environment variables to a subprocess.

davemfish commented:
This looks great to me. I really appreciate having those shell scripts and the new checklist to go along with them.
